pallets/dapp-staking-v3/src/lib.rs

```diff
@@ -1769,6 +1763,67 @@ pub mod pallet {
                 })
                 .collect::<Vec<_>>();
 
+            // 4.
+            // Iterate over configured tier and potential dApps.
+            // Each dApp will be assigned to the best possible tier if it satisfies the required condition,
+            // and tier capacity hasn't been filled yet.
+            let mut dapp_tiers = BTreeMap::new();
+            let mut tier_slots = BTreeMap::new();
+
+            let mut upper_bound = Balance::zero();
+            let mut rank_rewards = BoundedVec::<Balance, T::NumberOfTiers>::new();
+
+            for (tier_id, (tier_capacity, tier_threshold)) in tier_config
+                .slots_per_tier
+                .iter()
+                .zip(tier_config.tier_thresholds.iter())
+                .enumerate()
+            {
+                let lower_bound = tier_threshold.threshold();
+
+                // Iterate over dApps until one of two conditions has been met:
+                // 1. Tier has no more capacity
+                // 2. dApp doesn't satisfy the tier threshold (since they're sorted, none of the following dApps will satisfy the condition either)
+                for (dapp_id, staked_amount) in dapp_stakes
+                    .iter()
+                    .skip(dapp_tiers.len())
+                    .take_while(|(_, amount)| tier_threshold.is_satisfied(*amount))
+                    .take(*tier_capacity as usize)
+                {
+                    let rank = if T::RankingEnabled::get() {
+                        RankedTier::find_rank(lower_bound, upper_bound, *staked_amount)
+                    } else {
+                        0
+                    };
+                    tier_slots.insert(*dapp_id, RankedTier::new_saturated(tier_id as u8, rank));
+                }
+
+                // sum of all ranks for this tier
+                let ranks_sum = tier_slots
+                    .iter()
+                    .fold(0u32, |accum, (_, x)| accum.saturating_add(x.rank().into()));
```

Comment:
Minor optimization, but why not do it directly in the loop above?

Reply:
I just prefer functional. It works both ways. This just separate concerns